### PR TITLE
Add thread stepInto/stepOut/stepOver support

### PR DESCRIPTION
**Pull request by testforstephen**

### CODE REVIEW

**Review by andxu — APPROVED** [review]

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 1]
```diff
@@ -294,7 +292,17 @@ public DebugAdapter(Consumer<Events.DebugEvent> consumer) {
...
             subscription.dispose();
         });
-        
+        // If debuggee vm is still alive, then destroy debuggee vm forcibly.
+        if (this.debugSession.process().isAlive()) {
+            this.debugSession.terminate();
+        }
+        // Terminate eventHub thread.
+        try {
+            this.debugSession.eventHub().close();
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:301]: Event Hub should be closed when the VMDisconnectEvent is received. Otherwise, event hub will keep pulling events from event-queue and result in a VMDisconnectedException.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 2]
```diff
@@ -294,7 +292,17 @@ public DebugAdapter(Consumer<Events.DebugEvent> consumer) {
         this.eventSubscriptions.forEach(subscription -> {
             subscription.dispose();
         });
-        
+        // If debuggee vm is still alive, then destroy debuggee vm forcibly.
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:295]: Need to check the disconnect arguments to see the value of terminateDebuggee. If true, we need to call debugSession.terminate(). Otherwise, debugSession.detach() should be called and leave the target JVM running.
https://github.com/Microsoft/vscode-debugadapter-node/blob/master/protocol/src/debugProtocol.ts#L331

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 3]
```diff
@@ -294,7 +292,17 @@ public DebugAdapter(Consumer<Events.DebugEvent> consumer) {
...
+            this.debugSession.terminate();
+        }
+        // Terminate eventHub thread.
+        try {
+            this.debugSession.eventHub().close();
+        } catch (Exception e) {
+            // do nothing.
+        }
+        // Send a TerminatedEvent to indicate that the debugging of the debuggee has terminated.
         this.sendEvent(new Events.TerminatedEvent());
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:306]: This event needs to be send by receiving events from EventHub.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/internal/EventHub.java [hunk 4]
```diff
@@ -61,14 +61,18 @@ public void start(VirtualMachine vm) {
 
                     EventSet set = queue.remove();
 
-                    // Print JDI Events
-                    StringBuffer buf = new StringBuffer("\nJDI Event Set: {");
-                    for (Event event : set) {
-                        buf.append(event);
-                        buf.append(", ");
+                    try {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/internal/EventHub.java:64]: Suggest log the event in the for loop below by simply calling Login.logInfo on each event instead of visiting the whole event set again.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/internal/EventHub.java [hunk 5]
```diff
@@ -61,14 +61,18 @@ public void start(VirtualMachine vm) {
...
+                        buf.append("}\n");
+                        Logger.logInfo(buf.toString());
+                    } catch (VMDisconnectedException e) {
+                        // Do nothing.
                     }
-                    buf.append("}\n");
-                    Logger.logInfo(buf.toString());
 
                     boolean shouldResume = true;
                     for (Event event : set) {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/internal/EventHub.java:78]: Log the event here.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java [hunk 6]
```diff
@@ -464,18 +484,13 @@ private void handleEvent(DebugEvent debugEvent) {
...
-            Location bpLocation = bpEvent.location();
+        } else if (event instanceof BreakpointEvent || event instanceof StepEvent) {
             try {
-                Types.Source clientSource = this.convertDebuggerSourceToClient(bpLocation);
-                Events.StoppedEvent stopevent = new Events.StoppedEvent("breakpoint", clientSource,
-                        this.convertDebuggerLineToClient(bpLocation.lineNumber()), 0, "", bpThread.uniqueID());
-                this.stoppedEventsByThread.put(bpThread.uniqueID(), stopevent);
-                this.sendEvent(stopevent);
+                Events.StoppedEvent stopEvent = this.convertDebuggerStoppedEventToClient((LocatableEvent) event);
+                this.stoppedEventsByThread.put(stopEvent.threadId, stopEvent);
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java:490]: As discussed, we will not maintain internal state of the stopped threads but let VSCode handle the state management. Issue should be created at their side.
